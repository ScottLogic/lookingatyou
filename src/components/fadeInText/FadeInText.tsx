import React from 'react';
import { fadeInText } from '../../AppConstants';
import { normalise } from '../../utils/objectTracking/calculateFocus';
import './fadeInText.css';

interface IFadeInTextProps {
    text: string;
    show: boolean;
}

const FadeInText = React.memo((props: IFadeInTextProps) => {
    function renderSpans(text: string) {
        const textArray = text.split('');

        const combineWithDelays = (character: string) => ({
            character,
            delay: normalise(
                Math.random(),
                1,
                0,
                fadeInText.delayMax,
                fadeInText.delayMin,
            ),
        });

        return textArray.map(combineWithDelays).map(renderToSpan);
    }

    function renderToSpan(
        characterProps: { character: string; delay: number },
        index: number,
    ) {
        const transitionTime = normalise(
            Math.random(),
            1,
            0,
            fadeInText.transitionMax,
            fadeInText.transitionMin,
        );
        const style = {
            opacity: props.show ? 1 : 0,
            transition: `opacity ${transitionTime}ms`,
            transitionDelay: `${characterProps.delay}ms`,
            transitionTimingFunction: 'linear',
        };
        return (
            <span style={style} key={index}>
                {characterProps.character}
            </span>
        );
    }

    return <div className="revealText">{renderSpans(props.text)}</div>;
});

export default FadeInText;
