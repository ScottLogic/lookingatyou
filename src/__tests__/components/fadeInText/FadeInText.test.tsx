import { shallow } from 'enzyme';
import React from 'react';
import FadeInText from '../../../components/fadeInText/FadeInText';
describe('FadeInText', () => {
    const props = {
        text: 'test',
        show: true,
    };
    it('should render correctly', () => {
        const wrapper = shallow(<FadeInText {...props} />).debug();
        expect(wrapper).toMatchSnapshot();
    });
});
