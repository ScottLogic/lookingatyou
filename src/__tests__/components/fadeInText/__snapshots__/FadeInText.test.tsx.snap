// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FadeInText should render correctly 1`] = `
"<div className=\\"revealText\\">
  <span style={{...}}>
    t
  </span>
  <span style={{...}}>
    e
  </span>
  <span style={{...}}>
    s
  </span>
  <span style={{...}}>
    t
  </span>
</div>"
`;
